### PR TITLE
[SM-495] Access Policies - Individual Service Account - Project Tab

### PR DESCRIPTION
**Pull request by Thomas-Avery**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

The purpose of this PR is to create endpoints required to support the individual service account - projects tab for secrets manager.

## Code changes

- bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/CreateAccessPoliciesCommandTests.cs
src/Core/SecretsManager/Commands/AccessPolicies/Interfaces/ICreateAccessPoliciesCommand.cs

Refactor to just one public method to support new endpoints.

- bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/AccessPolicyRepository.cs
src/Core/SecretsManager/Repositories/IAccessPolicyRepository.cs

New method for fetch access policies by the service account grantee.

- src/Api/SecretsManager/Controllers/AccessPoliciesController.cs

Add in three new methods and some refactoring. 

`[HttpGet("/service-accounts/{id}/granted-policies")]`

Fetch access policies where the service account is the grantee

`[HttpPost("/service-accounts/{id}/granted-policies")]`

Create access policies where the service account is the grantee

`[HttpGet("/organizations/{id}/access-policies/projects/potential-grantees")]`

Fetch projects that could potentially be used in access policy creation. 

- src/Api/SecretsManager/Models/Request/GrantedAccessPolicyRequest.cs

New request model for creation endpoint. 

- src/Api/SecretsManager/Models/Response/AccessPolicyResponseModel.cs

Add granted project name for responses. 

- src/Api/SecretsManager/Models/Response/PotentialGranteeResponseModel.cs

Add projects as a potential target

- src/Core/SecretsManager/Repositories/IProjectRepository.cs
bitwarden_license/src/Commercial.Infrastructure.EntityFramework/SecretsManager/Repositories/ProjectRepository.cs

New method for fetching projects that the user has write permissions to. 

- test/Api.Test/SecretsManager/Controllers/AccessPoliciesControllerTests.cs
test/Api.IntegrationTest/SecretsManager/Controllers/AccessPoliciesControllerTest.cs
bitwarden_license/test/Commercial.Core.Test/SecretsManager/AccessPolicies/CreateAccessPoliciesCommandTests.cs

Update tests 

## Before you submit

- Please check for formatting errors (`dotnet format --verify-no-changes`) (required)
- If making database changes - make sure you also update Entity Framework queries and/or migrations
- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs [hunk 1]
```diff
@@ -58,9 +58,16 @@ public class AccessPoliciesController : Controller
     public async Task<ProjectAccessPoliciesResponseModel> CreateProjectAccessPoliciesAsync([FromRoute] Guid id,
         [FromBody] AccessPoliciesCreateRequest request)
     {
-        var userId = _userService.GetProperUserId(User).Value;
+        var project = await _projectRepository.GetByIdAsync(id);
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply src/Api/SecretsManager/Controllers/AccessPoliciesController.cs:61]: The check is performed in `GetAccessClientTypeAsync`. 

If we prefer to check in the controller itself, I can separate it all back out.

File: bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs [hunk 2]
```diff
@@ -83,6 +61,40 @@ private static void CheckForDistinctAccessPolicies(IReadOnlyCollection<BaseAcces
         }
     }
 
+    public async Task<IEnumerable<BaseAccessPolicy>> CreateManyAsync(List<BaseAccessPolicy> accessPolicies, Guid userId, AccessClientType accessType)
+    {
+        CheckForDistinctAccessPolicies(accessPolicies);
+        await CheckAccessPoliciesDoNotExistAsync(accessPolicies);
+        await CheckCanCreateAsync(accessPolicies, userId, accessType);
+        return await _accessPolicyRepository.CreateManyAsync(accessPolicies);
```
[the remark this answers is not in the document]
Comment (Thomas-Avery) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs:69]: The `CheckCanCreateAsync` we will be making N number of queries, where N is the number of granted objects involved in the creation request.

The
`[HttpPost("/projects/{id}/access-policies")]` and 
`[HttpPost("/service-accounts/{id}/access-policies")]` endpoints will have one query for this check.

The `[HttpPost("/service-accounts/{id}/granted-policies")]` endpoint, queries will be equal to the number of creation requests sent for this check.


For the `CheckAccessPoliciesDoNotExistAsync` this currently will be N number of queries, where N is the number of creation requests.

This check is where we should be able to optimize into one call. Currently, I'm struggling to get something to actually work.

[the remark this answers is not in the document]
Comment (Hinton) [reply bitwarden_license/src/Commercial.Core/SecretsManager/Commands/AccessPolicies/CreateAccessPoliciesCommand.cs:69]: We should probably make a follow up ticket for this and be satisfied with the constraint we've introduced.

File: src/Api/SecretsManager/Controllers/AccessPoliciesController.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Hinton) [reply src/Api/SecretsManager/Controllers/AccessPoliciesController.cs:67]: As long as we have the check we should be good.

**Review by Hinton — APPROVED** [review]

**Pull request merged**